### PR TITLE
feat(query) gRPC based communication between Query services

### PR DESCRIPTION
**Pull request by amolnayak311**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 
The communication between query services in multiple partitions currently uses the JSON based endpoint for data transfer. This JSON payload has no predefined schema and no natural mapping from the internal FiloDB data structures (``RangeVectors``). The format lacks flexibility, is error prone to changes and cumbersome to test locally. Also, current implementation does not support all aggregations like topk, bottomk across multiple partitions.

**New behavior :**

We will use gRPC for communication between partitions which will map FiloDB internal data structures to and back from proto format.

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/RemoteExec.scala [hunk 1]
```diff
@@ -95,6 +100,83 @@ trait RemoteExec extends LeafExecPlan with StrictLogging {
...
+
+  override def borrowChannel(endpointUrl: String): ManagedChannel = buildChannelFromEndpoint(endpointUrl)
+
+  override def returnChannel(channel: ManagedChannel): Unit = channel.shutdown()
+
+  override def shutdown(): Unit = {}
+}
+
+// TODO: Test for concurrency and performance. The handling for retrying after connection loss needs to be worked on
+class ReusableGRPCChannelManager extends BaseChannelManager {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/RemoteExec.scala:157]: who is responsible for keep-alive (or termination) of unused connections?

Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/RemoteExec.scala:157]: Thanks @vishramachandran this class is still not completely developed. I need to put in some work here. The simple implementation is what I used for current tests locally, we should use this implementation to manage channels. I will be working on this more in coming days.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 2]
```diff
@@ -9,7 +9,9 @@ import filodb.query._
 import filodb.query.LogicalPlan._
 import filodb.query.exec._
 
-case class PartitionAssignment(partitionName: String, endPoint: String, timeRange: TimeRange)
```
Comment (amolnayak311) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:12]: Rename ``endPoint`` to ``httpEndPoint``

File: query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                           queryContext: QueryContext,
+                           dispatcher: PlanDispatcher,
+                           dataset: DatasetRef) extends GrpcRemoteExec {
+    override def sendGrpcRequest(span: Span, requestTimeoutMs: Long)(implicit sched: Scheduler): Future[QueryResponse]
+    = Future {
+        val blockingStub = RemoteExecGrpc.newBlockingStub(channel)
+        import scala.collection.JavaConverters._
+        import filodb.query.ProtoConverters._
+        blockingStub.withDeadlineAfter(requestTimeoutMs, TimeUnit.MILLISECONDS)
+          .execStreaming(getGrpcRequest).asScala.toQueryResponse
```
Comment (amolnayak311) [inline query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala:71]: Make non blocking.

File: core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala [hunk 4]
```diff
@@ -39,12 +39,17 @@ import filodb.memory.format.{vectors => bv}
...
+ *                      NOTE: This value is not serialized over the wire in Kryo currently. See toSerializableTuple
+ *                      method for what gets serialized currently
  */
 final class RecordSchema(val columns: Seq[ColumnInfo],
                          // val hash: Int,
                          val partitionFieldStart: Option[Int] = None,
                          val predefinedKeys: Seq[String] = Nil,
-                         val brSchema: Map[Int, RecordSchema] = Map.empty) {
+                         val brSchema: Map[Int, RecordSchema] = Map.empty,
+                         val schemaVersion: Option[Int] = None) {
```
Comment (amolnayak311) [inline core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala:52]: Don't keep it as optional.

File: standalone/src/main/scala/filodb.standalone/FiloServer.scala [hunk 5]
```diff
@@ -65,6 +68,26 @@ class FiloServer(watcher: Option[ActorRef]) extends FilodbClusterNode {
     val singleton = cluster.clusterSingleton(role, watcher)
     filoHttpServer = new FiloHttpServer(cluster.system, cluster.settings)
     filoHttpServer.start(coordinatorActor, singleton, false, bootstrapper.getAkkaHttpRoute())
+    Thread.sleep(100)
```
Comment (amolnayak311) [inline standalone/src/main/scala/filodb.standalone/FiloServer.scala:71]: Add v2 related code to`` NewFiloServerMain``

File: query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    def getGrpcRequest: Request = {
+        val builder = Request.newBuilder()
+        builder.setQueryParams(promQlQueryParams.toProto)
+        builder.setPlannerParams(queryContext.plannerParams.copy(processMultiPartition = false).toProto)
+        builder.build()
+    }
+
+
+}
+case class PromQLGrpcRemoteExec(channel: Channel,
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala:63]: Typically ExecPlans are serializable. But `Channel` arg will create problems with serialization of `PromQLGrpcRemoteExec`.
Is it okay to treat as an exception? Or should we think about refactoring this code ?

File: standalone/src/main/scala/filodb.standalone/FiloServer.scala [hunk 7]
```diff
@@ -65,6 +68,25 @@ class FiloServer(watcher: Option[ActorRef]) extends FilodbClusterNode {
...
```
Comment (vishramachandran) [inline standalone/src/main/scala/filodb.standalone/FiloServer.scala:71]: Why was the sleep required ? The sleep with magic number can create unreliability

**Review by vishramachandran — APPROVED** [review]
Congrats on delivering the code changes for this long awaited capability.
Looks great! Thanks for the in-person code reviews and iterations. Looking forward for this to take fruit.

Left a couple of comments. Please address at your discretion.

File: coordinator/src/test/scala/filodb.coordinator/queryplanner/PlannerHierarchySpec.scala [hunk 8]
```diff
@@ -1912,6 +1928,233 @@ class PlannerHierarchySpec extends AnyFunSpec with Matchers with PlanValidationS
     }
   }
 
+
+  it ("should pushdown when leaf plans are split across partitions when one partition has gRPC QS") {
```
Comment (alextheimer) [inline coordinator/src/test/scala/filodb.coordinator/queryplanner/PlannerHierarchySpec.scala:1932]: Thank you for checking this 🙏

File: query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+trait GrpcRemoteExec extends LeafExecPlan {
+
+
+    def promQlQueryParams: PromQlQueryParams = queryContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+
+    // Since execute of overridden no one will invoke this
+    override def doExecute(source: ChunkSource, querySession: QuerySession)
+                          (implicit sched: Scheduler): ExecResult = ???
+
+    override def execute(source: ChunkSource,
```
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala:33]: You're already aware and the fix is out of the scope of this PR, but I'll call it out: overridden `execute`s won't automatically apply `RangeVectorTransformers`.

File: standalone/src/main/scala/filodb.standalone/NewFiloServerMain.scala [hunk 10]
```diff
@@ -54,4 +59,29 @@ object NewFiloServerMain extends StrictLogging {
         logger.error("Error occurred when initializing FiloDB server", e)
     }
   }
+
+  private def startGrpcServer(settings: FilodbSettings, coordinatorActor: ActorRef): Unit = {
+      if (settings.config.getBoolean("grpc.start-grpc-service")) {
+        val client = new LocalClient(coordinatorActor)
+        // TODO: Remove hardcoding
+        val dsRef = DatasetRef("prometheus")
```
Comment (alextheimer) [inline standalone/src/main/scala/filodb.standalone/NewFiloServerMain.scala:67]: Can these be made configurable in this PR?

**Review by alextheimer — APPROVED** [review]

File: standalone/src/main/scala/filodb.standalone/NewFiloServerMain.scala (hunk 10 above)
Comment (amolnayak311) [reply standalone/src/main/scala/filodb.standalone/NewFiloServerMain.scala:67]: I couldn't find a good existing property that I can use. I'll leave it for now if thats ok.

File: query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala (hunk 6 above)
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/PromQLGrpcRemoteExec.scala:78]: Made this class extend from ``RemoteExec`` for now but ``RemoteExec`` makes assumptions on HTTP as transport and ideally RemoteExec's subclass/trait should be ``HttpRemoteExec`` and Http code should go there. That cleanup can be done later.

**Pull request merged**